File: code/ex647.java
// DP

class Solution {
    public int countSubstrings(String s) {
        int n = s.length();
        int cnt = 0;
        
        // Define and initialize the dp array
        boolean[][] dp = new boolean[n][n];
        for (int i = 0; i < n; i++) {
            cnt++;
            dp[i][i] = true;
        }
        
        // Build dp array
        for (int l = 2; l <= n; l++) {
            for (int lo = 0; lo <= n - l; lo++) {
                int hi = lo + l - 1;
                if (s.charAt(lo) == s.charAt(hi)) {
                    if (dp[lo + 1][hi - 1] || lo + 1 >= hi - 1) {
                        cnt++;
                        dp[lo][hi] = true;
                    }
                }
            }
        }
        
        return cnt;
    }
}

File: code/ex647.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int n = s.length();
        
        int cnt = n;
        auto dp = vector<vector<bool>>(n, vector<bool>(n));
        for (int i = 0; i < n; i++)    dp[i][i] = 1;
        
        for (int l = 2; l <= n; l++) {
            for (int i = 0; i <= n - l; i++) {
                int j = i + l - 1;
                if (s[i] == s[j])    dp[i][j] = i + 1 <= j - 1 ? dp[i + 1][j - 1] : 1;
                if (dp[i][j])    cnt++;
            }
        }
        
        return cnt;
    }
};